---
title: "Get coordinates of 161 genes"
author: "Alexey Larionov"
date: "16 Feb 2021"
output:
  html_document:
    toc: true
    number_sections: true
    toc_float: 
      collapsed: false
      smooth_scroll: false
editor_options:
  chunk_output_type: console
---

# Summary

Get coordinates of 161 genes  

# Start section

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
options(width = 999)
```

```{r}

# Time stamp
Sys.time()

# Clenan-up
rm(list=ls())
graphics.off()

# Memory
gc()

# Options
options(stringsAsFactors=F)

# Files and folders
base_folder="/Users/alexey/Documents/mg/s2021/zhang_tests"
working_folder=file.path(base_folder,"s04_explore_gene_list")
setwd(working_folder)

# This database package is used to retrieve coordinates of human exones/genes
# The genes are given by entrez-ids
#BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")
library(TxDb.Hsapiens.UCSC.hg38.knownGene)

```

# Read 161 genes

```{r}

# Load data
load(file.path(working_folder,"s02_161_gene_names.RData"))

# Update folders (just in case)
base_folder="/Users/alexey/Documents/mg/s2021/zhang_tests"
working_folder=file.path(base_folder,"s04_explore_gene_list")

```

# Get genes coordinates

```{r}

# Connect to the database
b38.tx <- TxDb.Hsapiens.UCSC.hg38.knownGene

# Check connection
b38.tx
length(seqlevels(b38.tx))
seqlevels(b38.tx)[1:30]
columns(b38.tx)

# Default genes() setting prevent showing genes with multiple locations in genome
# i.e. located in canonical and in Alt loci (see ?genes)
genes_152.gr <- genes(b38.tx, 
                      filter=list(gene_id = genes_161.df$ENTREZID))
genes_152.gr

# If we suppress the dafult, we have all genes, but the output is in the LIST of genomic ranges
genes_161.grl <- genes(b38.tx, 
                       filter=list(gene_id = genes_161.df$ENTREZID), 
                       single.strand.genes.only=F)

# Example of a gene that is located on two "strands" i.e. in canonical chromosome and on an Alt loci
genes_161.grl[5] 

# Genes with Alt Loci: may need different processing to embrace all b38 complexity/richness
x <- names(which(lapply(genes_161.grl,length)>1)) # Entres ID
genes_161.df[genes_161.df$ENTREZID %in% x,]

# Set of canonical chromosomes for future filtering
# (to avoid gene versions annotated to Alt loci)
chromosomes <- c("chr1","chr2","chr3","chr4","chr5",
                 "chr6","chr7","chr8","chr9","chr10",
                 "chr11","chr12","chr13","chr14","chr15",
                 "chr16","chr17","chr18","chr19","chr20",
                 "chr21","chr22","chrX","chrY","chrM")

# Select gene coordinates 
# (use coordinates on canonical chromosomes only)
genes_161.gr <- genes(b38.tx, 
                       filter=list(gene_id = genes_161.df$ENTREZID, 
                                   tx_chrom = chromosomes))

# Sort by position
genes_161.gr <- sort(genes_161.gr)

# Check result
genes_161.gr

# Clean-up
rm(b38.tx, chromosomes, x, genes_152.gr, genes_161.grl)

```

# Add symbols

```{r}

# Sync records order in genes_161.df 
# with records order in genes_161.gr
head(genes_161.df)
genes_161.df <- genes_161.df[genes_161.gr$gene_id,]
head(genes_161.df)

# Add Symbols to GR object
genes_161.gr$symbol <- genes_161.df$SYMBOL

# Check result
genes_161.gr

# Clean-up
rm(genes_161.df)

```

# Save data

```{r}

save.image("s03_161_gene_coordinates.RData")

```

# Final section

```{r}

ls()
sessionInfo()
Sys.time()
gc()

```
