---
title: "Merge variants from text files and multisample VCF"
author: "Alexey Larionov"
date: "15 Apr 2021"
output:
  html_document:
    toc: true
    number_sections: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary

Combine data from TXT andf VCF SE:  

- Select and keep only variants overlapping between TXT and VCF  
- Append selected vannotations from VCF variants data to TXT genomic ranges  
- Use samples data from TXT (with updated col-names)  
- Use all matrices from both TXT (genotypes and scores) and VCF (gt, gt_add, dp, gq and ad)  
- Use metadata from both TXT and VCF  

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
options(width=999)
```

# Start section

```{r}

# Time
Sys.time()

# Clean up
rm(list=ls())
graphics.off()
gc()

# Options
options(stringsAsFactors = F)

# Working folders
# Files and folders
base_folder <- "/Users/alexey/Documents/mg/s2021/zhang_tests"
data_folder <- file.path(base_folder,"data","s12_merge_data_from_txt_and_vcf")
#dir.create(data_folder)
scripts_folder <- file.path(base_folder,"scripts","s12_merge_data_from_txt_and_vcf")
setwd(scripts_folder)

# Required libraries
library(dplyr)

```

# Load data

## Data from text files

```{r}

# Load
load(file.path(base_folder,"data/s10_add_vcf_coordinates/s01_add_vcf_coordinates.RData"))

# Restore folders
base_folder <- "/Users/alexey/Documents/mg/s2021/zhang_tests"
data_folder <- file.path(base_folder,"data","s12_merge_data_from_txt_and_vcf")
scripts_folder <- file.path(base_folder,"scripts","s12_merge_data_from_txt_and_vcf")

# Check SE
zheng.se

# Make row-names consistent with VCF data
rownames(zheng.se) <- rowRanges(zheng.se)$VCF_ID
zheng.se

#x.gr <- rowRanges(zheng.se)
#x.df <- rowData(zheng.se)

# Clean-up
rm(variants.df, functions_folder)

```

## Data from nultisample VCF

```{r}

# Load
load(file.path(base_folder,"data/s11_vep_annotated_multisample_vcf/s04_make_se.RData"))

# Restore folders
base_folder <- "/Users/alexey/Documents/mg/s2021/zhang_tests"
data_folder <- file.path(base_folder,"data","s12_merge_data_from_txt_and_vcf")
scripts_folder <- file.path(base_folder,"scripts","s12_merge_data_from_txt_and_vcf")

# Check SE
zheng_multisample_vcf.se

```

# Keep only shared variants

```{r}

# Identify shared variants
shared_ids_index <- rownames(zheng.se) %in% rownames(zheng_multisample_vcf.se)
base::table(shared_ids_index, useNA = "always")
shared_ids <- rownames(zheng.se)[shared_ids_index]

# Keep only shared variants
txt.se <- zheng.se[shared_ids,]
txt.se

vcf.se <- zheng_multisample_vcf.se[shared_ids,]
vcf.se

# Clean-up
rm(shared_ids_index,shared_ids,zheng.se,zheng_multisample_vcf.se)

```

# Prepare data for merged se

## Row data: variants info  

Add selected VCF annotations (ClinVar, VEP and other) to GRanges from TXT data  

Note using **mcols** and **values** to modify meta-data in GRanges  
https://kasperdanielhansen.github.io/genbioconductor/html/GenomicRanges_GRanges_Usage.html  

```{r}

# Extract variants data
txt_variants.gr <- rowRanges(txt.se)
vcf_variants.df <- rowData(vcf.se)

# Explore 
head(txt_variants.gr)[,1:2]
head(vcf_variants.df)[,1:5]

# Update colnames in TXT variants
colnames(mcols(txt_variants.gr))
txt_colnames <- colnames(mcols(txt_variants.gr))
txt_colnames <- paste("txt",txt_colnames,sep="_")
txt_colnames -> colnames(mcols(txt_variants.gr))
colnames(mcols(txt_variants.gr))

# Select annotations in VCF variants
colnames(vcf_variants.df)
selected_annotations <- c(
  "ID","CHROM","POS","END","REF","ALT",
  "QUAL","DP","DS","MULTIALLELIC",
  "CLNSIG","CLNREVSTAT","CLNDN",
  "vep_SYMBOL","vep_BIOTYPE","vep_VARIANT_CLASS",
  "vep_Consequence","vep_IMPACT",
  "vep_Existing_variation","vep_SOMATIC","vep_HGVSc","vep_HGVSp",
  "vep_Codons","vep_cDNA_position","vep_CDS_position","vep_CCDS",
  "vep_Amino_acids","vep_Protein_position",
  "vep_AF","vep_gnomAD_AF","vep_MAX_AF","vep_MAX_AF_POPS",
  "vep_MOTIF_NAME","vep_MOTIF_POS","vep_HIGH_INF_POS","vep_MOTIF_SCORE_CHANGE","vep_TRANSCRIPTION_FACTORS",
  "SIFT_call","SIFT_score","PolyPhen_call","PolyPhen_score","vep_CADD_PHRED","vep_CADD_RAW",
  "vep_NEAREST","vep_DOMAINS","vep_PUBMED")

selected_annotations.df <- vcf_variants.df[,selected_annotations]

colnames(selected_annotations.df) <- c(
  "vcf_ID","vcf_CHROM","vcf_POS","vcf_END","vcf_REF","vcf_ALT",
  "vcf_QUAL","vcf_DP","vcf_DS","vcf_MULTIALLELIC",
  "vcf_CLNSIG","vcf_CLNREVSTAT","vcf_CLNDN",
  "vep_SYMBOL","vep_BIOTYPE","vep_VARIANT_CLASS",
  "vep_Consequence","vep_IMPACT",
  "vep_Existing_variation","vep_SOMATIC","vep_HGVSc","vep_HGVSp",
  "vep_Codons","vep_cDNA_position","vep_CDS_position","vep_CCDS",
  "vep_Amino_acids","vep_Protein_position",
  "vep_AF","vep_gnomAD_AF","vep_MAX_AF","vep_MAX_AF_POPS",
  "vep_MOTIF_NAME","vep_MOTIF_POS","vep_HIGH_INF_POS","vep_MOTIF_SCORE_CHANGE","vep_TRANSCRIPTION_FACTORS",
  "vep_SIFT_call","vep_SIFT_score","vep_PolyPhen_call","vep_PolyPhen_score","vep_CADD_PHRED","vep_CADD_RAW",
  "vep_NEAREST","vep_DOMAINS","vep_PUBMED")

# Sanity check
sum(rownames(txt_variants.gr) != rownames(selected_annotations.df))

# Add annotations 
values(txt_variants.gr) <- data.frame(values(txt_variants.gr),selected_annotations.df)

# Rename (for clarity)
variants.gr <- txt_variants.gr

# Clean-up
rm(txt_colnames,vcf_variants.df,selected_annotations,selected_annotations.df,txt_variants.gr)

```

## Col data: samples info

- The colData should be a Data Frame  
- For now: just keep info from TXT data  
- Later more phenotype data could be added if needed  

```{r}

# Extract colData
txt_samples.df <- colData(txt.se)
vcf_samples.df <- colData(vcf.se)

# Explore
head(txt_samples.df)
head(vcf_samples.df)

# Sanity check
sum(txt_samples.df$id != vcf_samples.df$id)

# Add row-names (just in case)
rownames(txt_samples.df) <- txt_samples.df$id

# Update col-names
c("id","txt_file","study") -> colnames(txt_samples.df)

# Rename (for clarity)
samples.df <- txt_samples.df

# Clean-up
rm(vcf_samples.df,txt_samples.df)

```

## Metadata  

- Metadata is optional  
- If present, Metadata should be a list  
- Just in case, keep metadata from both TXT and VEP se  

```{r}

metadata.ls <- list(metadata(txt.se),metadata(vcf.se))
metadata.ls

```

## Assays: gt, dp, scores etc  

Keep all matrices from both TXT and VEP data   

Examples of working with summarised experiments:  
https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html  

```{r}

# txt
assays(txt.se)

gt_txt.mx <- assays(txt.se)$genotypes # Extracted oddly 
bass_scores.mx <- assays(txt.se)$scores

# vcf
assays(vcf.se)

gt_vcf.mx <- assays(vcf.se)$gt # Extracted oddly 
gt_add_vcf.mx <- assays(vcf.se)$gt_add
dp_vcf.mx <- assays(vcf.se)$dp
gq_vcf.mx <- assays(vcf.se)$gq
ad_vcf.mx <- assays(vcf.se)$ad # Extracted oddly 

# Clean-up
rm(txt.se,vcf.se)

```

# Pack merged data to a new SE

```{r}

zheng_merged.se <- SummarizedExperiment(
  assays=list(gt_txt=gt_txt.mx,
              bass_scores=bass_scores.mx,
              gt_vcf=gt_vcf.mx, 
              gt_add_vcf=gt_add_vcf.mx, 
              dp_vcf=dp_vcf.mx, 
              gq_vcf=gq_vcf.mx, 
              ad_vcf=ad_vcf.mx),
  rowRanges=variants.gr, 
  colData=samples.df, 
  metadata=metadata.ls)

# Check result
zheng_merged.se

assays(zheng_merged.se)$gt_txt[1:15,1:5] # Looks OK
assays(zheng_merged.se)$gt_vcf[1:15,1:5]
assays(zheng_merged.se)$gt_add_vcf[1:15,1:5]

# Clean-up
rm(samples.df,metadata.ls,variants.gr,
   gt_txt.mx,bass_scores.mx,
   gt_vcf.mx,gt_add_vcf.mx,dp_vcf.mx,gq_vcf.mx,ad_vcf.mx)

```

# Summarize and save

```{r}

ls()

save.image(file.path(data_folder,"s02_merge_variants_from_txt_and_vcf.RData"))

```

# Final section

```{r}

sessionInfo()
Sys.time()
gc()

```