---
title: "Add VCF-style coordinates to Zhang SE"
author: "Alexey Larionov"
date: "31 Mar 2021"
output:
  html_document:
    toc: true
    number_sections: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary

Add VCF-style coordinates to variants annotated in Annovar-style  

**SNP**

- VCF.Chr = Annovar.Chr  
- VCF.Pos = Annovar.Start  
- VCF.Ref = Annovar.Ref  
- VCF.Alt = Annovar.Alt  

**Deletion**

- VCF.Chr = Annovar.Chr  
- VCF.Pos = Annovar.Start-1 (because VCF adds one ancoring nucleotide)  
- VCF.Ref = Ref.genome[VCF.Pos, Annovar.End] (sequence from ref genome)  
- VC.Alt = Ref.genome[VCF.Pos] (ancoring nucleotide from ref genome)  

**Insertion**

- VCF.Chr = Annovar.Chr  
- VCF.Pos = Annovar.Start
- VCF.Ref = Ref.genome[Annovar.Start]  
- VC.Alt = VCF.Ref + Annovar.Alt

# Start section

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
options(width = 999)
```

```{r}

# Time stamp
Sys.time()

# Clenan-up
rm(list=ls())
graphics.off()

# Memory
gc()

# Options
options(stringsAsFactors=F)

# Files and folders
base_folder <- "/Users/alexey/Documents/mg/s2021/zhang_tests"
data_fiolder <- file.path(base_folder,"data","s10_add_vcf_coordinates")
#dir.create(data_fiolder)
scripts_fiolder <- file.path(base_folder,"scripts","s10_add_vcf_coordinates")
setwd(scripts_fiolder)

# Libraries
library(SummarizedExperiment)
library(BSgenome) # for reference genome

```

# Reference genome

Of course, the required "BSgenome.Hsapiens.UCSC.hg38" package should be pre-installed

```{r}

# Load required genome: b38 with "chr" prefixes
b38 <- getBSgenome("BSgenome.Hsapiens.UCSC.hg38")

# Sequence / chromosome names
b38@seqinfo

# Example of bases retrieval by coordinates
getSeq(b38,"chr1",start=257894,end=257897)

```

# Load data

```{r}

load(file.path(base_folder,"scripts","s08_read_zheng_to_se","s01_read_zheng_to_se.RData"))

rm(functions_fiolder)
base_folder <- "/Users/alexey/Documents/mg/s2021/zhang_tests"
data_fiolder <- file.path(base_folder,"data","s10_add_vcf_coordinates")
scripts_fiolder <- file.path(base_folder,"scripts","s10_add_vcf_coordinates")

# Explore SE
zheng.se
colData(zheng.se)
colnames(mcols(rowRanges(zheng.se)))

# Extract GRanges to data frame
variants.df <- as.data.frame(rowRanges(zheng.se))
dim(variants.df)
str(variants.df)
head(variants.df)
#View(varsiants.df)

```

# Add VCF-style coordinates

## Identify variant types

```{r}

# Select SNPs
snp <- variants.df$Ref %in% c("A","C","T","G") & 
       variants.df$Alt %in% c("A","C","T","G")
sum(snp)

# Select insertions
ins <- variants.df$Ref == "-"
sum(ins)

# Select deletions
del <- variants.df$Alt == "-"
sum(del)

# Count 
nrow(variants.df)
sum(snp)+sum(ins)+sum(del)

# Make var_type
var_type <- rep(NA,nrow(variants.df))
"snp" -> var_type[snp]
"ins" -> var_type[ins]
"del" -> var_type[del]
table(var_type, useNA="always")

# Add var_type column to data frame
variants.df <- data.frame(variants.df,var_type)

# Add var_type column to GRanges in SE
rowRanges(zheng.se)$var_type <- var_type
colnames(mcols(rowRanges(zheng.se)))

# Clean-up
rm(snp,ins,del,var_type)

```

## Calculate VCF-style coordinates

```{r}

# Initialize vectors for coordinates
VCF_Chr <- rep(NA,nrow(variants.df))
VCF_Pos <- rep(NA,nrow(variants.df))
VCF_Ref <- rep(NA,nrow(variants.df))
VCF_Alt <- rep(NA,nrow(variants.df))

# Loop over all variants
for(var_num in 1:nrow(variants.df)){
  
  # For debugging
  #var_num <- 1 # snp
  #var_num <- 7 # del
  #var_num <- 255 # ins
  
  # If SNP
  if(variants.df[var_num,"var_type"]=="snp")
  {
    VCF_Chr[var_num] <- as.character(variants.df[var_num,"seqnames"])
    VCF_Pos[var_num] <- variants.df[var_num,"start"]
    VCF_Ref[var_num] <- variants.df[var_num,"Ref"]
    VCF_Alt[var_num] <- variants.df[var_num,"Alt"]
  }
  
  # If Deletion
  if(variants.df[var_num,"var_type"]=="del")
  {
    VCF_Chr[var_num] <- as.character(variants.df[var_num,"seqnames"])
    VCF_Pos[var_num] <- variants.df[var_num,"start"] - 1
    VCF_Ref[var_num] <- as.character(getSeq(b38,VCF_Chr[var_num],
                               start=VCF_Pos[var_num],
                               end=variants.df[var_num,"end"]))
    VCF_Alt[var_num] <- as.character(getSeq(b38,VCF_Chr[var_num],
                               start=VCF_Pos[var_num],
                               end=VCF_Pos[var_num]))
  }
  
  # If Insertion
  if(variants.df[var_num,"var_type"]=="ins")
  {
    VCF_Chr[var_num] <- as.character(variants.df[var_num,"seqnames"])
    VCF_Pos[var_num] <- variants.df[var_num,"start"] - 1
    VCF_Ref[var_num] <- as.character(getSeq(b38,VCF_Chr[var_num],
                               start=VCF_Pos[var_num],
                               end=VCF_Pos[var_num]))
    VCF_Alt[var_num] <- paste0(VCF_Ref[var_num],
                               variants.df[var_num,"Alt"])
  }

} # Next variant

# Make joined Var ID
VCF_ID <- paste(VCF_Chr,VCF_Pos,VCF_Ref,VCF_Alt,sep="_")

# Add columns to the data frame
variants.df <- data.frame(variants.df,VCF_Chr,VCF_Pos,VCF_Ref,VCF_Alt,VCF_ID)

# Check updated data frame
variants.df[c(1,7,255),"VCF_ID",drop=F]
sum(is.na(variants.df[,c("VCF_Chr","VCF_Pos","VCF_Ref","VCF_Alt","VCF_ID")]))

# Add columns to the GRanges in SE
rowRanges(zheng.se)$VCF_Chr <- VCF_Chr
rowRanges(zheng.se)$VCF_Pos <- VCF_Pos
rowRanges(zheng.se)$VCF_Ref <- VCF_Ref
rowRanges(zheng.se)$VCF_Alt <- VCF_Alt
rowRanges(zheng.se)$VCF_ID <- VCF_ID

# Check updated GRanges in SE
colnames(mcols(rowRanges(zheng.se)))
rowRanges(zheng.se)[c(1,7,255),22:27]

# Clean-up
rm(var_num,VCF_Chr,VCF_Pos,VCF_Ref,VCF_Alt,VCF_ID,b38)

```

# Save data

```{r}

save.image(file.path(data_fiolder,"s01_add_vcf_coordinates.RData"))

```

# Final section

```{r}

ls()
sessionInfo()
Sys.time()
gc()

```
