---
title: "Combine two Zhang samples"
author: "Alexey Larionov"
date: "22 Feb 2021"
output:
  html_document:
    toc: true
    number_sections: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary

- Read two Zhang samples to GRanges  
- Preprocess  
  + Select variants on standard chromosomes  
  + Select variants overlapping with 161 genes  
  + Exclude common variants (MAF > 0.5)  
- Split x and y to subsets
  + Make Chr_Start_Ref_Alt key
  + Make subsets: unique x, y and overlap of x and y  
- Prepare merging
  + Make vector of variant annotations to keep in GRanges  
  + Make NA matrices for sample-specific variant annotations  
- Process unique x and y
  + Make GRanges for unique x and y  
  + Capture sample-specific annotations to the matrices  
- Porcess overlapping X and y  
  + Make GRange for intersect of x and y  
  + Capture sample-specific annotations to the matrices  
- Merge 
  + Merge unique and overlapping GRanges  
  + Reorder to match records in matrices  
- Cast to SummarisedExperiment class ?  

## notes

BASS Annotations in different text files can be different
so, selecting missenses and in-frames should be done in individual samples level

Then a minimal necessary set of metadata should be kept that are the same for any sample: 
e.g. only Ref and Alt ??? 

Use overlap type="equal"? 

subsetByOverlaps(gr12,setdiff(gr12,gr23))

re these euqal for metadata and for coordinates?

subsetByOverlaps(gr12,intersect(gr12,gr23)) 
subsetByOverlaps(gr23,intersect(gr12,gr23))

# Start section

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}
#options(width = 999)
```

```{r}

# Time stamp
Sys.time()

# Clenan-up
rm(list=ls())
graphics.off()

# Memory
gc()

# Options
options(stringsAsFactors=F)

# Files and folders
base_folder <- "/Users/alexey/Documents/mg/s2021/zhang_tests"
data_folder <- file.path(base_folder,"data","s06_combine_samples")
scripts_fiolder <- file.path(base_folder,"scripts","s06_combine_samples")
setwd(scripts_fiolder)

# Libraries
library(GenomicRanges)

```

# Read data

## 655 (of 667) genes

```{r}

# Load data
load(file.path(base_folder,"scripts","s04_gene_list_with_coordinates","s03_667_gene_coordinates.RData"))

# Remove unnecessary objects
rm(working_folder,absent_genes_9.df,genes_628.gr,genes_658.grl,genes_667.df,
   multi_genes_3.df, multi_genes_3.grl,multi_genes_30.df,multi_genes_30.grl)

# Update folders
base_folder <- "/Users/alexey/Documents/mg/s2021/zhang_tests"
data_folder <- file.path(base_folder,"data","s06_combine_samples")
scripts_fiolder <- file.path(base_folder,"scripts","s06_combine_samples")

# Check the genes list
summary(genes_655.gr)
seqinfo(genes_655.gr)
genes_655.gr

```

## Sample 1

```{r}

# Source file
sample1_file=file.path(base_folder,"data","d03_germline_annotated_files","RMSzhang_SJRHB020_hg38.bwa.QC.scores.filter.txt")

# Read to data frame
sample1.df <- read.table(sample1_file, 
                         header=T, sep = "\t", quote = "", na.strings="-9")

# Coerce data frame to GRanges
sample1.gr <- GRanges(sample1.df, seqinfo=Seqinfo(genome="hg38"))

# Check result
summary(sample1.gr)
seqinfo(sample1.gr)
sample1.gr[,1:2]

# Clean-up
rm(sample1_file,sample1.df)

```

## Sample 2

```{r}

# Source file
sample2_file=file.path(base_folder,"data","d03_germline_annotated_files","RMSzhang_SJRHB022_hg38.bwa.QC.scores.filter.txt")

# Read to data frame
sample2.df <- read.table(sample2_file, 
                         header=T, sep = "\t", quote = "", na.strings="-9")

# Coerse to GRanges
sample2.gr <- GRanges(sample2.df, seqinfo=Seqinfo(genome="hg38"))

# Check results
summary(sample2.gr)
seqinfo(sample2.gr)
sample2.gr[,1:2]

# Clean-up
rm(sample2_file,sample2.df)

```

## Make simple handles

```{r}

x <- sample1.gr
y <- sample2.gr
g <- genes_655.gr

```

# Preprocess

```{r}

# Keep only variants on standard chromosomes
x <- keepStandardChromosomes(x, pruning.mode="coarse")
x[,1:2]

y <- keepStandardChromosomes(y, pruning.mode="coarse")
y[,1:2]
  
# Overlap with the gene list
x <- subsetByOverlaps(x, g, type="any")
y <- subsetByOverlaps(y, g, type="any")
  
# Exclude common variants (MAF > 0.5)
x <- x[x$gnomAD_genome_ALL <= 0.05 | 
       x$gnomAD_genome_ALL >= 0.95 | 
       is.na(x$gnomAD_genome_ALL)]

y <- y[y$gnomAD_genome_ALL <= 0.05 | 
       y$gnomAD_genome_ALL >= 0.95 | 
       is.na(y$gnomAD_genome_ALL)]

# Clean-up
rm(g, genes_655.gr)

```

# Select sub-sets

Unique, Intersect and overlap:  
using key based on **chr**, **start**, **ref** and **alt**  

## Make x key

```{r}

# Select data for the key
x_chr <- as.character(x@seqnames)
x_start <- as.data.frame(x@ranges)$start
#x_end <- as.data.frame(x@ranges)$end
x_ref <- x$Ref
x_alt <- x$Alt

# Make and add key to GRanges
x_var_id <- paste(x_chr,x_start,x_ref,x_alt,sep="_")
x$var_id <- x_var_id

# Add key to GRanges names
# to facilitate easy selection of ranges by the key
names(x) <- x_var_id

# Check result
x[,c("var_id","Gene.refGene")]

# Clean-up
rm(x_chr,x_start,x_ref,x_alt,x_var_id)

```

## Make y key

```{r}

# Select data for key
y_chr <- as.character(y@seqnames)
y_start <- as.data.frame(y@ranges)$start
#y_end <- as.data.frame(y@ranges)$end
y_ref <- y$Ref
y_alt <- y$Alt

# Make and add key to GRanges
y_var_id <- paste(y_chr,y_start,y_ref,y_alt,sep="_")
y$var_id <- y_var_id

# Add key to GRanges names
# to facilitate easy selection of ranges by the key
names(y) <- y_var_id 

# Check result
y[,c("var_id","Gene.refGene")]

# Clean-up
rm(y_chr,y_start,y_ref,y_alt,y_var_id)

```

## Get subsets

```{r}

unique_x <- setdiff(x$var_id, y$var_id)
length(unique_x)

unique_y <- setdiff(y$var_id, x$var_id)
length(unique_y)

intersect_xy <- intersect(x$var_id, y$var_id)
length(intersect_xy)

union_xy <- union(x$var_id, y$var_id)
length(union_xy)

#length(c(unique_x,unique_y,intersect_xy)) == length(union_xy)

```

# Select annotations

## Stable annotations

These variant annotations should not change sample to sample, e.g. Ref, ALt, Gene name etc.  Such annotations will be kept within metadata of the merged GRanges object.  

```{r}

# Review available annotations
colnames(mcols(x))
#x.df <- as.data.frame(x)

# Select annotations to keep
stable_annotations <- c("var_id",
"Gene.refGene","Func.refGene","ExonicFunc.refGene",
"CLNSIG","CLNREVSTAT","CLNDN",
"avsnp150","HGVS_cDNA","HGVS_Prot",
"Ref","Alt",
"CDS_Pos","TranscriptSize",
"AA_Pos","AA1","AA2","ProteinSize","Interpro_domain",
"gnomAD_genome_ALL","ExAC_ALL")

```

## Unstable annotations

These variant annotations may change from sample to sample, such as Genotype or Drepth etc.  Such annotations will be captured and stored in additional matrices, syncronised with the GRanges object.  

Eventually this combination of matrices and GRanges resembles **SummarizedExperiment** class, which might be added here later for consistency with existent Bioconductor approaches:  

http://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html  

```{r}

# Unstable annotations

# Score
score.mx <- matrix(NA,ncol=2,nrow=length(union_xy))
rownames(score.mx) <- union_xy
colnames(score.mx) <- c("s1","s2")
head(score.mx)

#"TruncScore"
#"HomozygScore"
#"XAcorrect"

# Other matrices

# GT
gt.mx <- matrix(NA,ncol=2,nrow=length(union_xy))
rownames(gt.mx) <- union_xy
colnames(gt.mx) <- c("s1","s2")
head(gt.mx)

#"Depth"
#"AD"

```

# Process non-overlapping variants

## GRanges

```{r}

unique_x.gr <- x[unique_x,stable_annotations]
summary(unique_x.gr)

unique_y.gr <- y[unique_y,stable_annotations]
summary(unique_y.gr)

```

## Matrices

```{r}

score.mx[unique_x,"s1"] <- as.data.frame(x[unique_x,])$Score
gt.mx[unique_x,"s1"] <- as.data.frame(x[unique_x,])$GT

score.mx[unique_y,"s2"] <- as.data.frame(y[unique_y,])$Score
gt.mx[unique_y,"s2"] <- as.data.frame(y[unique_y,])$GT

```

# Process overlapping variants

## Granges

It may be simplifyed by removing the consistency check  

```{r}

# Get intersect ranges from both samples
intersect_x.gr <- x[intersect_xy,stable_annotations]
summary(intersect_x.gr)

intersect_y.gr <- y[intersect_xy,stable_annotations]
summary(intersect_y.gr)

# Check that intersect ranges from both samples are equal
if (!all(intersect_x.gr==intersect_y.gr)) {
  stop("GRanges shared elements differ", ok)
}

# Save the intersect
intersect_xy.gr <- intersect_x.gr

# Clean-up
rm(intersect_x.gr,intersect_y.gr)

```

## Matrices

```{r}

score.mx[unique_x,"s1"] <- as.data.frame(x[unique_x,])$Score
gt.mx[unique_x,"s1"] <- as.data.frame(x[unique_x,])$GT

score.mx[unique_y,"s2"] <- as.data.frame(y[unique_y,])$Score
gt.mx[unique_y,"s2"] <- as.data.frame(y[unique_y,])$GT

```

# Combine overlapping and non-overlapping granges

```{r}

# Combine granges
xy.gr <- c(unique_x.gr,unique_y.gr,intersect_xy.gr)

# Keep order consistent with the matrices
xy.gr <- xy.gr[union_xy]

# Check result
summary(xy.gr)
seqlevels(xy.gr)

# Compare with matrices
xy.gr[,1:2]
score.mx[1:5,]
gt.mx[1:5,]

```

# Save data

```{r}

save.image("s01_read_annotated_txt_to_granges.RData")

```

# Final section

```{r}

ls()
sessionInfo()
Sys.time()
gc()

```

# etc

```{r}
library(GenomicRanges)

gr12 <- GRanges(seqnames = c("chr1", "chr2"),
             ranges = IRanges(c(1, 11), c(11, 20)),
             seqinfo = Seqinfo(c("chr1", "chr2", "chr3")),
             score = c("a","b"))

gr23 <- GRanges(seqnames = c("chr2", "chr3"),
             ranges = IRanges(c(11, 21), c(20, 30)),
             seqinfo = Seqinfo(c("chr1", "chr2", "chr3")),
             score = c("b","c"))

gr12
gr23

BASS Annotations in different text files can be different

subsetByOverlaps(gr12,setdiff(gr12,gr23))

# Are these euqal for metadata and for coordinates?
subsetByOverlaps(gr12,intersect(gr12,gr23))
subsetByOverlaps(gr23,intersect(gr12,gr23))

subsetByOverlaps(gr23,setdiff(gr23,gr12))


u <- union(x,y)
?findOverlaps
queryHits(overlapsAny(u, x))

elementMetadata(u[overlapsAny(u, x)])$x = elementMetadata(gr1)[,1]

elementMetadata(u[overlapsAny(u, gr2)])$gr2 = elementMetadata(gr2)[,1]

subsetByOverlaps(gr23,setdiff(gr23,gr12))

x <- GRanges(seqnames = c("chr1", "chr2"),
               ranges = IRanges(c(1, 11), c(11, 20)))

y <- GRanges(seqnames = c("chr2", "chr3"),
               ranges = IRanges(c(11, 21), c(20, 30)))
x
y
union(x,y)

```


```{r}

install.packages("SparseSummarizedExperiment")

```


# Web examples

## Approach by Pete Haitch  

Its a package available on the author's page on GitHub.  
However, it has not been submitted to Bioconductor or CRAN.    
Also, it has not been developed for several years (since 2015).  

https://github.com/PeteHaitch/SparseSummarizedExperiment/blob/master/ideas/combining-SummarizedExperiment-objects.Rmd  

```{r eval=F}

setMethod("combine", c("GRanges", "GRanges"),
          function(x, y, ...) {
            
            if (length(y) == 0L) {
              return(x)
            } else if (length(x) == 0L) {
              return(y)
            }
            
            if (is.null(names(x)) || is.null(names(y))) {
              stop("'names' of 'x' and 'y' must be non-NULL")
            }
            shared_elements <- intersect(names(x), names(y))
            ok <- all.equal(x[shared_elements], y[shared_elements])
            if (!isTRUE(ok)) {
              stop("GRanges shared elements differ: ", ok)
            }
            c(x, y[setdiff(names(y), shared_elements)], ignore.mcols = FALSE)
          }
)

```

## Another example of merging GRanges

https://support.bioconductor.org/p/87075/  
